docs/src/userguide/wrapping_CPlusPlus.rst

```diff
@@ -457,43 +457,46 @@ that most users won't need, but for those that do a full example follows::
     cdef int raise_py_error()
     cdef int something_dangerous() except +raise_py_error
 
-If something_dangerous raises a C++ exception then raise_py_error will be
-called, which allows one to do custom C++ to Python error "translations." If
-raise_py_error does not actually raise an exception a RuntimeError will be
-raised. This approach may also be used to manage custom Python exceptions
-created using the Python C API. ::
-
-    # raising.pxd
-    cdef extern from "Python.h" nogil:
-        ctypedef struct PyObject
+If ``something_dangerous`` raises a C++ exception then ``raise_py_error`` will be
+called during exception handling (i.e. typically, in a ``catch`` clause).
+At that point, ``raise_py_error`` should retrieve the current exception object,
+either by rethrowing the current exception with ``throw;`` and immediately catching it,
+or via ``std::current_exception()``.  It can then set a Python exception via the ``PyErr_Set*()``
+functions of the Python C-API to implement a custom error translation from C++ to Python.
+See the example below. If  ``raise_py_error`` does not actually raise an exception, a
+``RuntimeError`` will be raised.  This approach may also be used to manage custom
+Python exceptions created using the Python C API. ::
+
+    from cpython.object cimport PyObject
 
     cdef extern from *:
         """
         #include <Python.h>
         #include <stdexcept>
         #include <ios>
 
-        PyObject *CustomLogicError;
-
-        void create_custom_exceptions() {
-            CustomLogicError = PyErr_NewException("raiser.CustomLogicError", NULL, NULL);
-        }
-
-        void custom_exception_handler() {
+        void cpp_handle_exception(PyObject *exc_type) {
+            if (PyErr_Occurred())
+                return; // let the latest Python exn pass through and ignore the current one
             try {
-                if (PyErr_Occurred()) {
-                    ; // let the latest Python exn pass through and ignore the current one
-                } else {
-                    throw;
-                }
+                throw;
             }  catch (const std::logic_error& exn) {
-                // Add mapping of std::logic_error -> CustomLogicError
-                PyErr_SetString(CustomLogicError, exn.what());
+                PyErr_SetString(exc_type, exn.what());
             } catch (...) {
                 PyErr_SetString(PyExc_RuntimeError, "Unknown exception");
             }
         }
+		"""
+
+        cdef void cpp_handle_exception(object)
+
+    class CustomLogicError(Exception): pass
 
+    cdef void translate_cpp_exception() except*:
```

Comment:
```suggestion
    cdef void translate_cpp_exception() except *:
        # Pass our custom Python exception types into the C++ handler,
        # in case it wants to raise them.
```